### PR TITLE
KMP classic algorithm and unit tests

### PR DESCRIPTION
**Pull request by jonathanmcchesney**
Story: #124

Added KMP algorithm to classics folder, along with a series of unit tests to bring the file coverage up to 100%. 
Updated file to keep in line with eslint rules. 
Added comments and time complexity of the KMP pattern matching algorithm.
Added example use case in file preamble.

### CODE REVIEW

File: src/_Classics_/knuth-morris-pratt/index.js [hunk 1]
```diff
@@ -0,0 +1,100 @@
+
+/* *
+* The time complexity of KMP algorithm is O(n) in the worst case
+* Example use case: Pattern = AABCAB Text = AAABACABAABCABAABCA
+* LPSArray = [ 0, 0, 1, 2, 3, 0 ]
```
Comment (TheSTL) [inline src/_Classics_/knuth-morris-pratt/index.js]: It should be LPSArray = [ 0, 1, 0, 0, 1, 0 ]

Comment (jonathanmcchesney) [reply src/_Classics_/knuth-morris-pratt/index.js]: Apologies, code results in this LPS: [ 0, 1, 0, 0, 1, 0 ]. I was using different inputs and forgot to swap out the LPS output for the new values I included in this example. Will update the PR with new test data.

Will update the example to show for: KMPSearch("ABCABCACA","AAABCBAABCABCACACABBCA");

Where: 
P = ABCABCACA
T = AAABCBAABCABCACACABBCA
lpsArray: [ 0, 0, 0, 1, 2, 3, 4, 0, 1 ]
Pattern Found at index 7

I feel this is a more useful example than the one I included above.

Comment (jonathanmcchesney) [reply src/_Classics_/knuth-morris-pratt/index.js]: PR has been updated with the new example, cheers.

File: src/_Classics_/knuth-morris-pratt/index.js [hunk 2]
```diff
@@ -0,0 +1,100 @@
+
+/* *
+* The time complexity of KMP algorithm is O(n) in the worst case
+* Example use case: Pattern = ABCABCACA Text = AAABCBAABCABCACACABBCA
+* LPSArray = [ 0, 0, 0, 1, 2, 3, 4, 0, 1 ]
```
Comment (TheSTL) [inline src/_Classics_/knuth-morris-pratt/index.js]: It should be LPSArray = [ 0, 1, 0, 0, 1, 0 ]

Comment (jonathanmcchesney) [reply src/_Classics_/knuth-morris-pratt/index.js]: Hi @TheSTL the pattern had been updated to ABCABCACA. I feel this is a more intuitive example. I believe [ 0, 0, 0, 1, 2, 3, 4, 0, 1 ] to be the correct LPS, if there are any more queries or amendments you would like to be made, I’ll updage the PR as appropriate. Cheers.

Comment (TheSTL) [reply src/_Classics_/knuth-morris-pratt/index.js]: I think this example is more understandable from previous one to compute LPS.

**Pull request merged**